File: web/test/read.test.ts
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { assert, expect, test } from "vitest";
import parseSchema from "../src/parseSchema";
import read from "../src/read";

const tmpDir = path.join(os.tmpdir(), "sdnx-read-test-" + Date.now());

function setupTestFiles(files: Record<string, string>): string[] {
	if (!fs.existsSync(tmpDir)) {
		fs.mkdirSync(tmpDir, { recursive: true });
	}
	const paths: string[] = [];
	for (const [name, content] of Object.entries(files)) {
		const filePath = path.join(tmpDir, name);
		const dir = path.dirname(filePath);
		if (!fs.existsSync(dir)) {
			fs.mkdirSync(dir, { recursive: true });
		}
		fs.writeFileSync(filePath, content, "utf-8");
		paths.push(filePath);
	}
	return paths;
}

function cleanupTestFiles(): void {
	if (fs.existsSync(tmpDir)) {
		fs.rmSync(tmpDir, { recursive: true, force: true });
	}
}

test.afterEach(() => {
	cleanupTestFiles();
});

test("read: successful with @schema directive", () => {
	const schema = `{ name: string, age: int }`;
	const data = `@schema("./schema.sdnx")
{ name: "Alice", age: 30 }`;

	const [_schemaPath, dataPath] = setupTestFiles({
		"schema.sdnx": schema,
		"data.sdn": data,
	});

	const result = read(dataPath);

	expect(result.ok).toBe(true);
	if (result.ok) {
		expect(result.data).toEqual({ name: "Alice", age: 30 });
	}
});

test("read: successful with explicit schema path", () => {
	const schema = `{ name: string, age: int }`;
	const data = `{ name: "Bob", age: 25 }`;

	const [schemaPath, dataPath] = setupTestFiles({
		"schema.sdnx": schema,
		"data.sdn": data,
	});

	const result = read(dataPath, schemaPath);

	expect(result.ok).toBe(true);
	if (result.ok) {
		expect(result.data).toEqual({ name: "Bob", age: 25 });
	}
});

test("read: successful with Schema object", () => {
	const schema = parseSchema(`{ name: string, age: int }`);
	assert(schema.ok);

	const data = `{ name: "Charlie", age: 35 }`;
	const [dataPath] = setupTestFiles({ "data.sdn": data });

	const result = read(dataPath, schema.data);

	expect(result.ok).toBe(true);
	if (result.ok) {
		expect(result.data).toEqual({ name: "Charlie", age: 35 });
	}
});

test("read: fails with data parse errors", () => {
	const schema = `{ name: string, age: int }`;
	const data = `@schema("./schema.sdnx")
{ name: "Alice", age: }`;

	const [_schemaPath, dataPath] = setupTestFiles({
		"schema.sdnx": schema,
		"data.sdn": data,
	});

	const result = read(dataPath);

	expect(result.ok).toBe(false);
	if (!result.ok) {
		expect(result.schemaErrors.length).toBeGreaterThan(0);
		expect(result.dataErrors).toHaveLength(0);
		expect(result.checkErrors).toHaveLength(0);
		expect(result.schemaErrors[0].message).toBeDefined();
	}
});

test("read: fails with schema parse errors", () => {
	const schema = `{ name: string, age: }`;
	const data = `@schema("./schema.sdnx")
{ name: "Alice", age: 30 }`;

	const [_schemaPath, dataPath] = setupTestFiles({
		"schema.sdnx": schema,
		"data.sdn": data,
	});

	const result = read(dataPath);

	expect(result.ok).toBe(false);
	if (!result.ok) {
		expect(result.schemaErrors.length).toBeGreaterThan(0);
		expect(result.dataErrors).toHaveLength(0);
		expect(result.checkErrors).toHaveLength(0);
	}
});

test("read: fails with validation errors", () => {
	const schema = `{ name: string, age: int min(18) }`;
	const data = `@schema("./schema.sdnx")
{ name: "Alice", age: 15 }`;

	const [_schemaPath, dataPath] = setupTestFiles({
		"schema.sdnx": schema,
		"data.sdn": data,
	});

	const result = read(dataPath);

	expect(result.ok).toBe(false);
	if (!result.ok) {
		expect(result.schemaErrors).toHaveLength(0);
		expect(result.dataErrors).toHaveLength(0);
		expect(result.checkErrors.length).toBeGreaterThan(0);
		expect(result.checkErrors[0].message).toContain("least");
	}
});

test("read: throws error when file not found", () => {
	expect(() => read("/nonexistent/path/to/file.sdn")).toThrow("File not found");
});

test("read: throws error when @schema directive missing and schema not provided", () => {
	const data = `{ name: "Alice", age: 30 }`;
	const [dataPath] = setupTestFiles({ "data.sdn": data });

	expect(() => read(dataPath)).toThrow("Schema required");
});

test("read: resolves relative schema path correctly", () => {
	const schema = `{ name: string }`;
	const data = `@schema("./schema.sdnx")
{ name: "Alice" }`;

	const [_schemaPath, dataPath] = setupTestFiles({
		"schema.sdnx": schema,
		"data.sdn": data,
	});

	const result = read(dataPath);

	expect(result.ok).toBe(true);
	if (result.ok) {
		expect(result.data).toEqual({ name: "Alice" });
	}
});

test("read: handles nested schema path", () => {
	const schema = `{ name: string }`;
	const data = `@schema("./schemas/schema.sdnx")
{ name: "Alice" }`;

	const paths = setupTestFiles({
		"schemas/schema.sdnx": schema,
		"data.sdn": data,
	});

	const result = read(paths[1]);

	expect(result.ok).toBe(true);
	if (result.ok) {
		expect(result.data).toEqual({ name: "Alice" });
	}
});

test("read: handles complex nested data", () => {
	const schema = `{
	name: string,
	age: int,
	address: { street: string, city: string },
	tags: [string]
}`;

	const data = `@schema("./schema.sdnx")
{
	name: "Alice",
	age: 30,
	address: { street: "123 Main St", city: "NYC" },
	tags: ["developer", "engineer"]
}`;

	const [_schemaPath, dataPath] = setupTestFiles({
		"schema.sdnx": schema,
		"data.sdn": data,
	});

	const result = read(dataPath);

	expect(result.ok).toBe(true);
	if (result.ok) {
		expect(result.data).toEqual({
			name: "Alice",
			age: 30,
			address: { street: "123 Main St", city: "NYC" },
			tags: ["developer", "engineer"],
		});
	}
});

test("read: includes line and char info in parse errors", () => {
	const schema = `{ name: string }`;
	const data = `@schema("./schema.sdnx")
{ name: "Alice",
age: invalid
}`;

	const [_schemaPath, dataPath] = setupTestFiles({
		"schema.sdnx": schema,
		"data.sdn": data,
	});

	const result = read(dataPath);

	expect(result.ok).toBe(false);
	if (!result.ok) {
		expect(result.schemaErrors.length).toBeGreaterThan(0);
		const error = result.schemaErrors[0];
		expect(error.line).toBeDefined();
		expect(error.char).toBeDefined();
		expect(error.index).toBeDefined();
		expect(error.length).toBeDefined();
		expect(error.message).toBeDefined();
	}
});

test("read: handles empty data file", () => {
	const schema = `{ name: string }`;
	const data = `@schema("./schema.sdnx")
{}`;

	const [_schemaPath, dataPath] = setupTestFiles({
		"schema.sdnx": schema,
		"data.sdn": data,
	});

	const result = read(dataPath);

	expect(result.ok).toBe(false);
	if (!result.ok) {
		expect(result.checkErrors.length).toBeGreaterThan(0);
	}
});

test("read: handles schema with union types", () => {
	const schema = `{ value: int | string }`;
	const data = `@schema("./schema.sdnx")
{ value: 42 }`;

	const [_schemaPath, dataPath] = setupTestFiles({
		"schema.sdnx": schema,
		"data.sdn": data,
	});

	const result = read(dataPath);

	expect(result.ok).toBe(true);
	if (result.ok) {
		expect(result.data).toEqual({ value: 42 });
	}
});

test("read: handles schema with array of objects", () => {
	const schema = `{ users: [{ name: string, age: int }] }`;
	const data = `@schema("./schema.sdnx")
{
	users: [
		{ name: "Alice", age: 30 },
		{ name: "Bob", age: 25 }
	]
}`;

	const [_schemaPath, dataPath] = setupTestFiles({
		"schema.sdnx": schema,
		"data.sdn": data,
	});

	const result = read(dataPath);

	expect(result.ok).toBe(true);
	if (result.ok) {
		expect(result.data).toEqual({
			users: [
				{ name: "Alice", age: 30 },
				{ name: "Bob", age: 25 },
			],
		});
	}
});

test("read: handles file path from cwd", () => {
	const schema = `{ name: string }`;
	const data = `@schema("./schema.sdnx")
{ name: "Alice" }`;

	const [_schemaPath, _dataPath] = setupTestFiles({
		"schema.sdnx": schema,
		"data.sdn": data,
	});

	const oldCwd = process.cwd();
	process.chdir(tmpDir);

	try {
		const result = read("data.sdn");

		expect(result.ok).toBe(true);
		if (result.ok) {
			expect(result.data).toEqual({ name: "Alice" });
		}
	} finally {
		process.chdir(oldCwd);
	}
});

File: zig/test/read.test.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

const sdn = @import("sdn");
const read_mod = sdn.read_mod;
const read = sdn.read;
const parse = sdn.parse;
const parseSchema = sdn.parseSchema;
const ParseResult = parse.ParseResult;
const SchemaParseResult = parseSchema.SchemaParseResult;
const Schema = sdn.parseSchema_mod.Schema;

const test_io = std.testing.io;
const test_allocator = std.testing.allocator;

var tmp_dir: std.Io.Dir = undefined;
var tmp_dir_path: []const u8 = undefined;

const cwd = std.Io.Dir.cwd();

fn setupTmpDir(_: Allocator) !void {
    tmp_dir = try cwd.createDirPathOpen(std.testing.io, "zig-test-tmp", .{});
    tmp_dir_path = "zig-test-tmp";
}

fn cleanupTmpDir(_: Allocator) void {
    if (tmp_dir_path.len > 0) {
        tmp_dir.deleteTree(std.testing.io, tmp_dir_path) catch {};
        tmp_dir.close(std.testing.io);
        cwd.deleteTree(std.testing.io, tmp_dir_path) catch {};
        tmp_dir_path = "";
    }
}

fn writeFile(rel_path: []const u8, content: []const u8) !void {
    // Use tmp_dir (the test temp directory) instead of cwd
    // Try to write directly first
    tmp_dir.writeFile(std.testing.io, .{ .sub_path = rel_path, .data = content }) catch |err| {
        if (err == error.FileNotFound) {
            // Directory doesn't exist, create it
            const dir_path = std.fs.path.dirname(rel_path) orelse ".";
            try tmp_dir.createDirPath(std.testing.io, dir_path);
            // Now try writing again
            try tmp_dir.writeFile(std.testing.io, .{ .sub_path = rel_path, .data = content });
            return;
        }
        return err;
    };
}

test "read: successful with @schema directive" {
    try setupTmpDir(test_allocator);
    defer cleanupTmpDir(test_allocator);

    const schema_content = "{ name: string, age: int }";
    const data_content =
        \\@schema("./schema.sdnx")
        \\{ name: "Alice", age: 30 }
    ;

    try writeFile("schema.sdnx", schema_content);
    try writeFile("data.sdn", data_content);

    const data_path = try std.fs.path.join(test_allocator, &.{ tmp_dir_path, "data.sdn" });
    defer test_allocator.free(data_path);

    var result = try read_mod.read(test_io, test_allocator, data_path, null);
    defer result.deinit();

    try std.testing.expect(result.ok);
    try std.testing.expect(result.data != null);
    if (result.data) |d| {
        try std.testing.expect(d.object.get("name") != null);
        try std.testing.expect(d.object.get("age") != null);
        try std.testing.expectEqual(@as(i64, 30), d.object.get("age").?.int);
    }
}

test "read: successful with explicit schema path" {
    try setupTmpDir(test_allocator);
    defer cleanupTmpDir(test_allocator);

    const schema_content = "{ name: string, age: int }";
    const data_content = "{ name: \"Bob\", age: 25 }";

    try writeFile("schema.sdnx", schema_content);
    try writeFile("data.sdn", data_content);

    const schema_path = try std.fs.path.join(test_allocator, &.{ tmp_dir_path, "schema.sdnx" });
    defer test_allocator.free(schema_path);
    const data_path = try std.fs.path.join(test_allocator, &.{ tmp_dir_path, "data.sdn" });
    defer test_allocator.free(data_path);

    // For this test, we'll parse the schema first
    const schema_contents = try std.Io.Dir.cwd().readFileAlloc(test_io, schema_path, test_allocator, .unlimited);
    defer test_allocator.free(schema_contents);
    var schema_parsed = parseSchema(test_allocator, schema_contents);
    defer schema_parsed.deinit();

    var result = try read_mod.read(test_io, test_allocator, data_path, &schema_parsed.schema.?);
    defer result.deinit();

    try std.testing.expect(result.ok);
    try std.testing.expect(result.data != null);
}

test "read: successful with Schema object" {
    try setupTmpDir(test_allocator);
    defer cleanupTmpDir(test_allocator);

    const schema_content = "{ name: string, age: int }";
    var schema_parsed = parseSchema(test_allocator, schema_content);
    defer schema_parsed.deinit();

    const data_content = "{ name: \"Charlie\", age: 35 }";
    try writeFile("data.sdn", data_content);

    const data_path = try std.fs.path.join(test_allocator, &.{ tmp_dir_path, "data.sdn" });
    defer test_allocator.free(data_path);

    var result = try read_mod.read(test_io, test_allocator, data_path, &schema_parsed.schema.?);
    defer result.deinit();

    try std.testing.expect(result.ok);
    try std.testing.expect(result.data != null);
}

test "read: fails with data parse errors" {
    try setupTmpDir(test_allocator);
    defer cleanupTmpDir(test_allocator);

    const schema_content = "{ name: string, age: int }";
    const data_content =
        \\@schema("./schema.sdnx")
        \\{ name: "Alice", age: }
    ;

    try writeFile("schema.sdnx", schema_content);
    try writeFile("data.sdn", data_content);

    const data_path = try std.fs.path.join(test_allocator, &.{ tmp_dir_path, "data.sdn" });
    defer test_allocator.free(data_path);

    var result = try read_mod.read(test_io, test_allocator, data_path, null);
    defer result.deinit();

    try std.testing.expect(!result.ok);
    try std.testing.expect(result.schemaErrors.items.len == 0);
    try std.testing.expect(result.dataErrors.items.len > 0);
    try std.testing.expect(result.checkErrors.items.len == 0);
}

test "read: fails with schema parse errors" {
    try setupTmpDir(test_allocator);
    defer cleanupTmpDir(test_allocator);

    const schema_content = "{ name: string, age: }";
    const data_content =
        \\@schema("./schema.sdnx")
        \\{ name: "Alice", age: 30 }
    ;

    try writeFile("schema.sdnx", schema_content);
    try writeFile("data.sdn", data_content);

    const data_path = try std.fs.path.join(test_allocator, &.{ tmp_dir_path, "data.sdn" });
    defer test_allocator.free(data_path);

    var result = try read_mod.read(test_io, test_allocator, data_path, null);
    defer result.deinit();

    try std.testing.expect(!result.ok);
    try std.testing.expect(result.schemaErrors.items.len > 0);
    try std.testing.expect(result.dataErrors.items.len == 0);
    try std.testing.expect(result.checkErrors.items.len == 0);
}

test "read: fails with validation errors" {
    try setupTmpDir(test_allocator);
    defer cleanupTmpDir(test_allocator);

    const schema_content = "{ name: string, age: int min(18) }";
    const data_content =
        \\@schema("./schema.sdnx")
        \\{ name: "Alice", age: 15 }
    ;

    try writeFile("schema.sdnx", schema_content);
    try writeFile("data.sdn", data_content);

    const data_path = try std.fs.path.join(test_allocator, &.{ tmp_dir_path, "data.sdn" });
    defer test_allocator.free(data_path);

    var result = try read_mod.read(test_io, test_allocator, data_path, null);
    defer result.deinit();

    try std.testing.expect(!result.ok);
    try std.testing.expect(result.schemaErrors.items.len == 0);
    try std.testing.expect(result.dataErrors.items.len == 0);
    try std.testing.expect(result.checkErrors.items.len > 0);
}

test "read: throws error when file not found" {
    const result_not_found = read_mod.read(test_io, test_allocator, "/nonexistent/path/to/file.sdn", null);
    try std.testing.expectError(error.FileNotFound, result_not_found);
}

test "read: resolves relative schema path correctly" {
    try setupTmpDir(test_allocator);
    defer cleanupTmpDir(test_allocator);

    const schema_content = "{ name: string }";
    const data_content =
        \\@schema("./schema.sdnx")
        \\{ name: "Alice" }
    ;

    try writeFile("schema.sdnx", schema_content);
    try writeFile("data.sdn", data_content);

    const data_path = try std.fs.path.join(test_allocator, &.{ tmp_dir_path, "data.sdn" });
    defer test_allocator.free(data_path);

    var result = try read_mod.read(test_io, test_allocator, data_path, null);
    defer result.deinit();

    try std.testing.expect(result.ok);
    try std.testing.expect(result.data != null);
}

test "read: handles nested schema path" {
    try setupTmpDir(test_allocator);
    defer cleanupTmpDir(test_allocator);

    const schema_content = "{ name: string }";
    const data_content =
        \\@schema("./schemas/schema.sdnx")
        \\{ name: "Alice" }
    ;

    try writeFile("schemas/schema.sdnx", schema_content);
    try writeFile("data.sdn", data_content);

    const data_path = try std.fs.path.join(test_allocator, &.{ tmp_dir_path, "data.sdn" });
    defer test_allocator.free(data_path);

    var result = try read_mod.read(test_io, test_allocator, data_path, null);
    defer result.deinit();

    try std.testing.expect(result.ok);
    try std.testing.expect(result.data != null);
}

test "read: handles complex nested data" {
    try setupTmpDir(test_allocator);
    defer cleanupTmpDir(test_allocator);

    const schema_content =
        \\{
        \\name: string,
        \\age: int,
        \\address: { street: string, city: string },
        \\tags: [string]
        \\}
    ;
    const data_content =
        \\@schema("./schema.sdnx")
        \\{
        \\name: "Alice",
        \\age: 30,
        \\address: { street: "123 Main St", city: "NYC" },
        \\tags: ["developer", "engineer"]
        \\}
    ;

    try writeFile("schema.sdnx", schema_content);
    try writeFile("data.sdn", data_content);

    const data_path = try std.fs.path.join(test_allocator, &.{ tmp_dir_path, "data.sdn" });
    defer test_allocator.free(data_path);

    var result = try read_mod.read(test_io, test_allocator, data_path, null);
    defer result.deinit();

    try std.testing.expect(result.ok);
    try std.testing.expect(result.data != null);
}

test "read: includes line and char info in parse errors" {
    try setupTmpDir(test_allocator);
    defer cleanupTmpDir(test_allocator);

    const schema_content = "{ name: string }";
    const data_content =
        \\@schema("./schema.sdnx")
        \\{ name: "Alice",
        \\age: invalid
        \\}
    ;

    try writeFile("schema.sdnx", schema_content);
    try writeFile("data.sdn", data_content);

    const data_path = try std.fs.path.join(test_allocator, &.{ tmp_dir_path, "data.sdn" });
    defer test_allocator.free(data_path);

    var result = try read_mod.read(test_io, test_allocator, data_path, null);
    defer result.deinit();

    try std.testing.expect(!result.ok);
    try std.testing.expect(result.dataErrors.items.len > 0);
    const err = result.dataErrors.items[0];
    try std.testing.expect(err.line.len > 0);
    try std.testing.expect(err.message.len > 0);
    try std.testing.expect(err.index > 0);
}

test "read: handles empty data file" {
    try setupTmpDir(test_allocator);
    defer cleanupTmpDir(test_allocator);

    const schema_content = "{ name: string }";
    const data_content =
        \\@schema("./schema.sdnx")
        \\{}
    ;

    try writeFile("schema.sdnx", schema_content);
    try writeFile("data.sdn", data_content);

    const data_path = try std.fs.path.join(test_allocator, &.{ tmp_dir_path, "data.sdn" });
    defer test_allocator.free(data_path);

    var result = try read_mod.read(test_io, test_allocator, data_path, null);
    defer result.deinit();

    try std.testing.expect(!result.ok);
    try std.testing.expect(result.checkErrors.items.len > 0);
}

test "read: handles schema with union types" {
    try setupTmpDir(test_allocator);
    defer cleanupTmpDir(test_allocator);

    const schema_content = "{ value: int | string }";
    const data_content =
        \\@schema("./schema.sdnx")
        \\{ value: 42 }
    ;

    try writeFile("schema.sdnx", schema_content);
    try writeFile("data.sdn", data_content);

    const data_path = try std.fs.path.join(test_allocator, &.{ tmp_dir_path, "data.sdn" });
    defer test_allocator.free(data_path);

    var result = try read_mod.read(test_io, test_allocator, data_path, null);
    defer result.deinit();

    try std.testing.expect(result.ok);
    try std.testing.expect(result.data != null);
}

test "read: handles schema with array of objects" {
    try setupTmpDir(test_allocator);
    defer cleanupTmpDir(test_allocator);

    const schema_content = "{ users: [{ name: string, age: int }] }";
    const data_content =
        \\@schema("./schema.sdnx")
        \\{
        \\users: [
        \\{ name: "Alice", age: 30 },
        \\{ name: "Bob", age: 25 }
        \\]
        \\}
    ;

    try writeFile("schema.sdnx", schema_content);
    try writeFile("data.sdn", data_content);

    const data_path = try std.fs.path.join(test_allocator, &.{ tmp_dir_path, "data.sdn" });
    defer test_allocator.free(data_path);

    var result = try read_mod.read(test_io, test_allocator, data_path, null);
    defer result.deinit();

    try std.testing.expect(result.ok);
    try std.testing.expect(result.data != null);
}
